File: 0047-permutations-ii/0047-permutations-ii.cpp
#include <vector>
#include <algorithm>

class Solution {
public:
    std::vector<std::vector<int>> permuteUnique(std::vector<int>& nums) {
        std::vector<std::vector<int>> result;
        std::sort(nums.begin(), nums.end()); // Sort the array to handle duplicates
        std::vector<bool> used(nums.size(), false);
        std::vector<int> current;
        backtrack(nums, used, current, result);
        return result;
    }

private:
    void backtrack(const std::vector<int>& nums, std::vector<bool>& used, std::vector<int>& current, std::vector<std::vector<int>>& result) {
        if (current.size() == nums.size()) {
            result.push_back(current);
            return;
        }
        for (int i = 0; i < nums.size(); ++i) {
            if (used[i] || (i > 0 && nums[i] == nums[i - 1] && !used[i - 1])) continue;
            used[i] = true;
            current.push_back(nums[i]);
            backtrack(nums, used, current, result);
            used[i] = false;
            current.pop_back();
        }
    }
};

File: 0047-permutations-ii/0047-permutations-ii.rs
impl Solution {
    pub fn permute_unique(nums: Vec<i32>) -> Vec<Vec<i32>> {
        let mut nums = nums;
        nums.sort(); // Sort the array to handle duplicates
        let mut result = Vec::new();
        let mut used = vec![false; nums.len()];
        let mut current = Vec::new();
        Self::backtrack(&nums, &mut used, &mut current, &mut result);
        result
    }

    fn backtrack(nums: &Vec<i32>, used: &mut Vec<bool>, current: &mut Vec<i32>, result: &mut Vec<Vec<i32>>) {
        if current.len() == nums.len() {
            result.push(current.clone());
            return;
        }
        for i in 0..nums.len() {
            if used[i] || (i > 0 && nums[i] == nums[i - 1] && !used[i - 1]) {
                continue;
            }
            used[i] = true;
            current.push(nums[i]);
            Self::backtrack(nums, used, current, result);
            used[i] = false;
            current.pop();
        }
    }
}
